File: test/f90/leafNumber.f90
MODULE leafNumberModule
    USE crop2mlModules
    IMPLICIT NONE

CONTAINS

    SUBROUTINE Calculate_LeafNumber(deltaTT,phyllochron,hasFlagLeafLiguleAppeared,&
            switchMaize, atip,leaf_tip_emerg,k_bl, nlim, leafNumber,&
            cumulTTPhenoMaizeAtEmergence, cumulTT,phase, ntip)			
        !inputs
        INTEGER , INTENT(IN):: hasFlagLeafLiguleAppeared, switchMaize
        REAL, INTENT(IN):: deltaTT, atip, leaf_tip_emerg, k_bl, &
            nlim, cumulTT,phase
        !outputs
        REAL, INTENT(INOUT):: leafNumber, cumulTTPhenoMaizeAtEmergence, phyllochron
        REAL, INTENT(OUT) ::  ntip
        REAL :: leafNumber1, abl, nextstartExpTT, tt_lim_lip, tt_bl, bbl, nexttipTT

        leafNumber1 =leafNumber
        ntip=0
        IF ((phase .EQ. 1) .AND. (cumulTTPhenoMaizeAtEmergence .EQ. 0)) THEN
            cumulTTPhenoMaizeAtEmergence = cumulTT
        END IF
        IF ((phase .GE. 1) .AND. (phase .LT. 4)) THEN
            IF (hasFlagLeafLiguleAppeared .EQ. 0) THEN
                IF (switchMaize .EQ. 0) THEN
                    IF (phyllochron .EQ. 0.0) THEN
                        phyllochron = 0.0000001
                    END IF
                    leafNumber = leafNumber1 + MIN(deltaTT / phyllochron, 0.999)
                    ntip=0.0
                ELSE
                    IF (leafNumber1 .LT. leaf_tip_emerg) THEN
                        leafNumber = leaf_tip_emerg
                    ELSE
                        nextstartExpTT = 0
                        nexttipTT = ((leafNumber1 + 1) - leaf_tip_emerg) / atip + cumulTTPhenoMaizeAtEmergence
                        abl = k_bl * atip
                        tt_lim_lip = ((nlim - leaf_tip_emerg) / atip) + cumulTTPhenoMaizeAtEmergence
                        bbl = nlim - (abl * tt_lim_lip)
                        tt_bl = ((leafNumber1 + 1) - bbl) / abl
                        IF (tt_bl .GT. nexttipTT) THEN
                            nextstartExpTT = nexttipTT
                        ELSE
                            nextstartExpTT = tt_bl
                        ENDIF
                        IF (cumulTT .GE. nextstartExpTT) THEN
                            leafNumber = leafNumber1 + 1
                        ELSE
                            leafNumber = leafNumber1
                        END IF
                    END IF
                    ntip = atip * (cumulTT - cumulTTPhenoMaizeAtEmergence) + leaf_tip_emerg
                END IF
            END IF
        END IF
    END SUBROUTINE
END MODULE

File: src/cs/LeafNumber.cs
public class LeafNumber
{
    public double leafNumber;
    public double ntip;
    public LeafNumber(double _leafNumber,double _ntip)
    {
        leafNumber=_leafNumber;
        ntip=_ntip;
    }
}

public static class Estimation_LeafNumber
{
    public static LeafNumber CalculateLeafNumber(double deltaTT,double phyllochron,int hasFlagLeafLiguleAppeared,int switchMaize,double atip,double leaf_tip_emerg,double k_bl,double nlim,double leafNumber,double cumulTTPhenoMaizeAtEmergence,double cumulTT,double phase)
    {


/*
     CalculateLeafNumber Model

    Author: 
    Reference: Modeling development phase in the 
                Wheat Simulation Model SiriusQuality.
                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    Instituton: INRA Montpellier
    Abstract: calculate leaf number. LeafNumber increase is caped at one more leaf per day
    
*/
        double ntip;

        double leafNumber1, nextstartExpTT, nexttipTT, abl, tt_lim_lip, bbl, tt_bl;
        ntip = 0;
        leafNumber1 =leafNumber;		
        if (phase == 1 && cumulTTPhenoMaizeAtEmergence == 0)
        {
        	cumulTTPhenoMaizeAtEmergence = cumulTT;			
        }		
        if (phase >= 1 && phase< 4)
        {
        	if (hasFlagLeafLiguleAppeared==0)//sowingToAnthesis
        	{
        		if (switchMaize==0)
        		{
        			if (phyllochron == 0.0)
        			{
        				phyllochron = 0.0000001;		
        			}
        			leafNumber = leafNumber1 + Math.Min(deltaTT / phyllochron, 0.999);
        		}
        		else
        		{		
        			if (leafNumber1 < leaf_tip_emerg)
        			{
        				leafNumber = leaf_tip_emerg;
        			}
        			else
        			{		
        				nextstartExpTT = 0;
        				nexttipTT = ((leafNumber1 + 1) - leaf_tip_emerg) / atip + cumulTTPhenoMaizeAtEmergence;
        				abl = k_bl * atip;
        				tt_lim_lip = ((nlim - leaf_tip_emerg) / atip) + cumulTTPhenoMaizeAtEmergence;
        				bbl = nlim - (abl * tt_lim_lip);
        				tt_bl = ((leafNumber1 + 1) - bbl) / abl;
        				if (tt_bl > nexttipTT)
        				{
        					nextstartExpTT = nexttipTT;
        				}
        				else
        				{
        					nextstartExpTT = tt_bl;
        				}
        				if (cumulTT>= nextstartExpTT)
        				{
        					leafNumber = leafNumber1 + 1;
        				}
        				else
        				{
        					leafNumber = leafNumber1;
        				}
        			}
        			ntip = atip * (cumulTT - cumulTTPhenoMaizeAtEmergence) + leaf_tip_emerg;
        		}
        	}
        }
        return new LeafNumber(leafNumber,ntip);
    }
}

